echo-model/src/main/java/com/netflix/spinnaker/echo/model/Trigger.java

```diff
@@ -86,6 +92,23 @@ public String toString() {
   List<String> expectedArtifactIds;
   Map<String, ?> lastSuccessfulExecution;
 
+  // this is set after deserialization, not in the json representation
+  @NonFinal
+  Pipeline parent;
+
+  public String getIdWithFallback() {
+    if (id != null) {
+      return id;
+    }
+
+    if (parent == null) {
+      log.warn("Trigger with unknown parent: {}", this);
+      return UUID.nameUUIDFromBytes(this.toString().getBytes()).toString();
+    }
+
+    return UUID.nameUUIDFromBytes((parent.getId() + "/" + this.toString()).getBytes()).toString();
```

Comment:
core piece: generating a fallback id if there isn't a hardcoded one, based on the parent pipeline plus value of this trigger

Reply:
Not sure I'm a huge fan of having both `id` and `getIdWithFallback` ... see below for a possible workaround.

Presumably you never actually want to call `id` directly?

Reply:
Indeed, we never want to call `id` directly. Thanks for the suggestion, let me massage it in!

Reply:
Massaging complete in da7d74f. I had to rework quite a few tests, since we are now generating copies of pipelines and triggers, but I think it is much cleaner now!